core/trino-main/src/main/java/io/trino/memory/ClusterMemoryManager.java

```diff
@@ -289,15 +304,15 @@ private synchronized void callOomKiller(Iterable<QueryExecution> runningQueries)
     @GuardedBy("this")
     private boolean isLastKillTargetGone(Iterable<QueryExecution> runningQueries)
     {
-        if (lastKillTarget == null) {
+        if (lastKillTarget.isEmpty()) {
             return true;
         }
 
-        if (lastKillTarget.isWholeQuery()) {
-            return isQueryGone(lastKillTarget.getQuery());
+        if (lastKillTarget.get().isWholeQuery()) {
+            return isQueryGone(lastKillTarget.get().getQuery());
         }
 
-        return areTasksGone(lastKillTarget.getTasks(), runningQueries);
+        return areTasksGone(lastKillTarget.get().getTasks(), runningQueries);
```

Comment:
Ideally the implementation of this method should be based on the information about the memory pool, as the decision to kill tasks is based on that view. What do you think?

Reply:
Technically it does not matter much I think. As we coordinator learns that task is dead only after it is actually killed on the worker. I can change that but I think it is low-prio cleanup and will not update this PR.

Reply:
This may technically introduce another inconsistency. On workers tasks are first being transitioned to the `FAILED` state and the cleanup of active operators happens asynchronously. Thus it is possible for a worker to report that a task is done, yet it may take some time to cleanup the memory pool. Hard to say how much does it matter in practice. It feels safer though. The `isQueryGone` is also implemented based on the `ClusterMemoryPool` which is based on the `MemoryInfo`. I wonder if it makes sense to make it consistent.

Reply:
Should be ok now. Though I did not test it yet. Will take a look tomorrow as I do not think it is covered by automation.

Reply:
PTAL at last commit.